README.md
# pytest-patterns

[![Build Status](https://travis-ci.org/smarie/pytest-patterns.svg?branch=master)](https://travis-ci.org/smarie/pytest-patterns) [![Documentation](https://img.shields.io/badge/docs-latest-blue.svg)](https://smarie.github.io/pytest-patterns/)

A couple of examples showing how to use core `pytest` mechanisms and existing plugins (no additional hooks or hacks) to solve real-world problems. In other words:
 
    "pytest for humans ;)"

**This is the readme for developers.** The documentation for users is available here: [https://smarie.github.io/pytest-patterns/](https://smarie.github.io/pytest-patterns/)


## Want to contribute ?

Contributions are welcome ! Simply fork this project on github, commit your contributions, and create pull requests.

Here is a non-exhaustive list of interesting open topics: [https://github.com/smarie/pytest-patterns/issues](https://github.com/smarie/pytest-patterns/issues)

## Running the tests

This project uses `pytest`.

```bash
pytest -v pytest_patterns/tests/
```

You may need to install requirements for setup beforehand, using 

```bash
pip install -r ci_tools/requirements-test.txt
```

## Generating the documentation page

This project uses `mkdocs` to generate its documentation page. Therefore building a local copy of the doc page may be done using:

```bash
mkdocs build -f docs/mkdocs.yml
```

You may need to install requirements for doc beforehand, using 

```bash
pip install -r ci_tools/requirements-doc.txt
```

## Generating the test reports

The following commands generate the html test report and the associated badge. 

```bash
pytest --junitxml=junit.xml -v pytest_patterns/tests/
ant -f ci_tools/generate-junit-html.xml
python ci_tools/generate-junit-badge.py
```

### Merging pull requests with edits - memo

Ax explained in github ('get commandline instructions'):

```bash
git checkout -b <git_name>-<feature_branch> master
git pull https://github.com/<git_name>/pytest-patterns.git <feature_branch> --no-commit --ff-only
```

if the second step does not work, do a normal auto-merge (do not use **rebase**!):

```bash
git pull https://github.com/<git_name>/pytest-patterns.git <feature_branch> --no-commit
```

Finally review the changes, possibly perform some modifications, and commit.


docs/changelog.md
# Changelog

### 0.3.0 - Update for latest pytest-cases

Data science example simplified now that `pytest-cases` 2.0.3 is available.

### 0.2.0 - Simplified datascience benchmark example 

Simplified datascience benchmark example thanks to usage of pytest harvest 1.2.0

### 0.1.1 - Fixed import error in example

### 0.1.0 - First public version


pytest_patterns/data_science_benchmark/test_polyfit.py
from pathlib import Path

import numpy as np
from pytest_cases import fixture, parametrize_with_cases
import logging

# logging configuration
exec_log = logging.getLogger('algo')
logs_dir = Path(__file__).parent / "logs"


@fixture(autouse=True)
def configure_logging(request, caplog):
    """ Set log file name same as test name, and set log level. You could change the format here too """
    log_file = logs_dir / ("%s.log" % request.node.name)
    request.config.pluginmanager.get_plugin("logging-plugin").set_log_path(log_file)
    caplog.set_level(logging.INFO)


@fixture
@parametrize_with_cases("algo", cases='.challengers_polyfit', prefix='algo_')
def challenger(algo):
    """ A fixutre collecting all challengers from `challengers_polyfit.py` """
    # (optional setup code here)
    yield algo
    # (optional teardown code here)


@fixture(scope='session')
@parametrize_with_cases("data", cases='.datasets_polyfit', prefix='data_')
def dataset(data):
    """ A fixture collecting all datasets from `datasets_polyfit.py`.
    Note: we use "scope=session" so that this method is called only once per case.
    This ensures that each file is read once.
    """
    # (optional setup code here)
    yield data
    # (optional teardown code here)


def test_poly_fit(challenger, dataset, results_bag):
    """ Evaluation protocol.
    Applies the `challenger` on the provided `dataset`, and stores the model accuracy (cv-rmse) in the results_bag
    """

    # Fit the model
    exec_log.info("fitting model")
    challenger.fit(dataset.x, dataset.y)
    results_bag.model = challenger

    # Use the model to perform predictions
    exec_log.info("predicting")
    predictions = challenger.predict(dataset.x)

    # Evaluate the prediction error
    exec_log.info("evaluating error")
    cvrmse = np.sqrt(np.mean((predictions-dataset.y)**2)) / np.mean(dataset.y)
    print("Relative error (cv-rmse) is: %.2f%%" % (cvrmse * 100))
    results_bag.cvrmse = cvrmse


# ------------- To create the final benchmark table ------------
def test_synthesis(module_results_df):
    """
    Creates the benchmark synthesis table
    Note: we could do this at many other places (hook, teardown of a session-scope fixture...)
    as explained in `pytest-harvest` plugin
    """
    # ----------- (1) `module_results_df` contains the raw (12 rows) table -----------
    # rename columns and only keep useful information
    module_results_df = rename_with_checks(module_results_df, columns={'challenger_param': 'degree',
                                                                       'dataset_param': 'dataset'})
    module_results_df['challenger'] = module_results_df['model'].map(str)  # only keep the string representation
    module_results_df = module_results_df[['dataset', 'challenger', 'degree', 'status', 'duration_ms', 'cvrmse']]

    # write to csv
    module_results_df.to_csv("polyfit_bench_results.csv", sep=';', decimal=',')

    # pretty-print (requires tabulate)
    try:
        from tabulate import tabulate
        print("\n" + tabulate(module_results_df, headers='keys', tablefmt='pipe'))
    except ImportError:
        pass

    # ----------- (2) graphical synthesis: bar chart (requires matplotlib)------------
    try:
        import matplotlib.pyplot as plt

        # convert all to categorical so that we can pivot
        module_results_df = module_results_df.apply(lambda s: s.astype("category") if s.dtype == 'object' else s)

        cvrmse_df = module_results_df[['dataset', 'challenger', 'cvrmse']].pivot(index='dataset',
                                                                                 columns='challenger',
                                                                                 values='cvrmse')

        ax = cvrmse_df.plot.bar()
        ax.set_ylabel("cvrmse")
        plt.xticks(plt.xticks()[0], plt.xticks()[1], rotation=30, ha='right')
        plt.subplots_adjust(left=0.20, bottom=0.25)
        print("Close the plots to continue...")
        plt.show()
    except ImportError:
        pass

    # ----------- (3) summarizing the results further - by challenger --------------
    summary_df = module_results_df[['degree', 'duration_ms', 'cvrmse']].groupby('degree', axis=0).agg({
        'duration_ms': ['mean', 'std'],
        'cvrmse': ['mean', 'std']
    })
    # pretty-print (requires tabulate)
    try:
        print("\n" + tabulate(summary_df, headers='keys'))
    except NameError:
        pass


def rename_with_checks(df, columns, **kwargs):
    """
    Same than df.rename(columns=<columns>, **kwargs) but checks that columns exist before executing.
    """
    missing = set(columns.keys()).difference(set(df.columns))
    if len(missing) > 0:
        raise ValueError("Missing columns: %s" % missing)

    return df.rename(columns=columns, **kwargs)


pytest_patterns/data_science_benchmark/test_polyfit_cases.py
from os import listdir, path

import numpy as np

from pytest_cases import cases_generator


def generate_anscombes_dataset(i):
    """ Returns the I-th dataset in the Anscombe's quartet """
    if i == 1:
        x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]
        y = [8.04, 6.95, 7.58, 8.81, 8.33, 9.96, 7.24, 4.26, 10.84, 4.82, 5.68]
    elif i == 2:
        x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]
        y = [9.14, 8.14, 8.74, 8.77, 9.26, 8.10, 6.13, 3.10, 9.13, 7.26, 4.74]
    elif i == 3:
        x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]
        y = [7.46, 6.77, 12.74, 7.11, 7.81, 8.84, 6.08, 5.39, 8.15, 6.42, 5.73]
    elif i == 4:
        x = [8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 19.0, 8.0, 8.0, 8.0]
        y = [6.58, 5.76, 7.71, 8.84, 8.47, 7.04, 5.25, 12.50, 5.56, 7.91, 6.89]
    else:
        raise ValueError("Anscombes quartet contains, well... 4 datasets. "
                         "Invalid i=%s" % i)

    return np.array(x, dtype=float), np.array(y, dtype=float)


@cases_generator("Anscombe's quartet {i}", i=range(1, 5))
def case_anscombe(i):
    """ This is where we can both provide the dataset AND the expected results/errors (TODO expected accuracy?) """

    # get data
    x, y = generate_anscombes_dataset(i)

    return x, y


THIS_DIR = path.dirname(__file__)
DATASETS_DIR = path.join(THIS_DIR, 'datasets')
ALL_DATASET_FILES = [(DATASETS_DIR, file_name) for file_name in listdir(DATASETS_DIR)]


@cases_generator("Data file '{csv_file_path[1]}'", csv_file_path=ALL_DATASET_FILES)
def case_file(csv_file_path):
    """ """
    my_data = np.genfromtxt(path.join(*csv_file_path), delimiter=',', names=True)
    x = my_data['x']
    y = my_data['y']

    return x, y


ci_tools/github_release.py
# a clone of the ruby example https://gist.github.com/valeriomazzeo/5491aee76f758f7352e2e6611ce87ec1
import os
from os import path

import re

import click
from click import Path
from github import Github, UnknownObjectException
# from valid8 import validate  not compliant with python 2.7


@click.command()
@click.option('-u', '--user', help='GitHub username')
@click.option('-p', '--pwd', help='GitHub password')
@click.option('-s', '--secret', help='GitHub access token')
@click.option('-r', '--repo-slug', help='Repo slug. i.e.: apple/swift')
@click.option('-cf', '--changelog-file', help='Changelog file path')
@click.option('-d', '--doc-url', help='Documentation url')
@click.option('-df', '--data-file', help='Data file to upload', type=Path(exists=True, file_okay=True, dir_okay=False,
                                                                          resolve_path=True))
@click.argument('tag')
def create_or_update_release(user, pwd, secret, repo_slug, changelog_file, doc_url, data_file, tag):
    """
    Creates or updates (TODO)
    a github release corresponding to git tag <TAG>.
    """
    # 1- AUTHENTICATION
    if user is not None and secret is None:
        # using username and password
        # validate('user', user, instance_of=str)
        assert isinstance(user, str)
        # validate('pwd', pwd, instance_of=str)
        assert isinstance(pwd, str)
        g = Github(user, pwd)
    elif user is None and secret is not None:
        # or using an access token
        # validate('secret', secret, instance_of=str)
        assert isinstance(secret, str)
        g = Github(secret)
    else:
        raise ValueError("You should either provide username/password OR an access token")
    click.echo("Logged in as {user_name}".format(user_name=g.get_user()))

    # 2- CHANGELOG VALIDATION
    regex_pattern = "[\s\S]*[\n][#]+[\s]*(?P<title>[\S ]*%s[\S ]*)[\n]+(?P<body>[\s\S]*?)[\n]*(\n#|$)" % re.escape(tag)
    changelog_section = re.compile(regex_pattern)
    if changelog_file is not None:
        # validate('changelog_file', changelog_file, custom=os.path.exists,
        #          help_msg="changelog file should be a valid file path")
        assert os.path.exists(changelog_file), "changelog file should be a valid file path"
        with open(changelog_file) as f:
            contents = f.read()

        match = changelog_section.match(contents).groupdict()
        if match is None or len(match) != 2:
            raise ValueError("Unable to find changelog section matching regexp pattern in changelog file.")
        else:
            title = match['title']
            message = match['body']
    else:
        title = tag
        message = ''

    # append footer if doc url is provided
    message += "\n\nSee [documentation page](%s) for details." % doc_url

    # 3- REPOSITORY EXPLORATION
    # validate('repo_slug', repo_slug, instance_of=str, min_len=1, help_msg="repo_slug should be a non-empty string")
    assert isinstance(repo_slug, str) and len(repo_slug) > 0, "repo_slug should be a non-empty string"
    repo = g.get_repo(repo_slug)

    # -- Is there a tag with that name ?
    try:
        tag_ref = repo.get_git_ref("tags/" + tag)
    except UnknownObjectException:
        raise ValueError("No tag with name %s exists in repository %s" % (tag, repo.name))

    # -- Is there already a release with that tag name ?
    click.echo("Checking if release %s already exists in repository %s" % (tag, repo.name))
    try:
        release = repo.get_release(tag)
        if release is not None:
            raise ValueError("Release %s already exists in repository %s. Please set overwrite to True if you wish to "
                             "update the release (Not yet supported)" % (tag, repo.name))
    except UnknownObjectException:
        # Release does not exist: we can safely create it.
        click.echo("Creating release %s on repo: %s" % (tag, repo.name))
        click.echo("Release title: '%s'" % title)
        click.echo("Release message:\n--\n%s\n--\n" % message)
        repo.create_git_release(tag=tag, name=title,
                                message=message,
                                draft=False, prerelease=False)

        # add the asset file if needed
        if data_file is not None:
            release = None
            while release is None:
                release = repo.get_release(tag)
            release.upload_asset(path=data_file, label=path.split(data_file)[1], content_type="application/gzip")

        # --- Memo ---
        # release.target_commitish  # 'master'
        # release.tag_name    # '0.5.0'
        # release.title       # 'First public release'
        # release.body        # markdown body
        # release.draft       # False
        # release.prerelease  # False
        # #
        # release.author
        # release.created_at  # datetime.datetime(2018, 11, 9, 17, 49, 56)
        # release.published_at  # datetime.datetime(2018, 11, 9, 20, 11, 10)
        # release.last_modified  # None
        # #
        # release.id           # 13928525
        # release.etag         # 'W/"dfab7a13086d1b44fe290d5d04125124"'
        # release.url          # 'https://api.github.com/repos/smarie/python-pytest-harvest/releases/13928525'
        # release.html_url     # 'https://github.com/smarie/python-pytest-harvest/releases/tag/0.5.0'
        # release.tarball_url  # 'https://api.github.com/repos/smarie/python-pytest-harvest/tarball/0.5.0'
        # release.zipball_url  # 'https://api.github.com/repos/smarie/python-pytest-harvest/zipball/0.5.0'
        # release.upload_url   # 'https://uploads.github.com/repos/smarie/python-pytest-harvest/releases/13928525/assets{?name,label}'


if __name__ == '__main__':
    create_or_update_release()


pytest_patterns/data_science_benchmark/challengers_polyfit.py
import numpy as np

from pytest_cases import parametrize


class BenchmarkChallenger(object):
    """ Represents the API that a challenger should implement to enter the benchmark """
    def fit(self, x, y):
        pass

    def predict(self, x):
        pass


class PolyFitChallenger(BenchmarkChallenger):
    """ A benchmark challenger implementation relying on np.polyfit """
    def __init__(self, degree):
        self.coefs = None
        self.degree = degree

    def __str__(self):
        return "NumpyPolyfit(degree=%i)" % self.degree

    def fit(self, x, y):
        self.coefs = np.polyfit(x, y, deg=self.degree)

    def predict(self, x):
        all_x_powers = np.c_[[x ** d for d in range(self.degree, -1, -1)]]
        predictions = self.coefs.dot(all_x_powers)
        return predictions


@parametrize(degree=[1, 2])
def algo_polyfit(degree):
    """ The two challengers based on polyfit, to be injected in the benchmark. """
    return PolyFitChallenger(degree=degree)


pytest_patterns/data_science_benchmark/datasets_polyfit.py
from collections import namedtuple
from pathlib import Path

import numpy as np
from pytest_cases import parametrize


Dataset = namedtuple("Dataset", ('x', 'y'))
"""A minimal structure representing a dataset"""


# ------------- example data created by scripts --------------
@parametrize(id=range(1, 5))
def data_anscombes_quartet(id):
    """ Generate one case for each of the Anscombe's quartet """
    if id == 1:
        x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]
        y = [8.04, 6.95, 7.58, 8.81, 8.33, 9.96, 7.24, 4.26, 10.84, 4.82, 5.68]
    elif id == 2:
        x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]
        y = [9.14, 8.14, 8.74, 8.77, 9.26, 8.10, 6.13, 3.10, 9.13, 7.26, 4.74]
    elif id == 3:
        x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]
        y = [7.46, 6.77, 12.74, 7.11, 7.81, 8.84, 6.08, 5.39, 8.15, 6.42, 5.73]
    elif id == 4:
        x = [8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 8.0, 19.0, 8.0, 8.0, 8.0]
        y = [6.58, 5.76, 7.71, 8.84, 8.47, 7.04, 5.25, 12.50, 5.56, 7.91, 6.89]
    else:
        raise ValueError("Anscombes quartet contains, well... 4 datasets. "
                         "Invalid i=%s" % id)

    x = np.array(x, dtype=float)
    y = np.array(y, dtype=float)
    return Dataset(x, y)


# -------------- example data created from files --------------
datasets_dir = Path(__file__).parent / 'datasets'
all_csv_files = [datasets_dir / file_name for file_name in datasets_dir.glob('*.csv')]


@parametrize(csv_file_path=all_csv_files, idgen=lambda csv_file_path: csv_file_path.stem)
def data_csvfile(csv_file_path):
    """ Generates one case per file in the datasets/ folder """
    my_data = np.genfromtxt(csv_file_path, delimiter=',', names=True)
    return Dataset(x=my_data['x'], y=my_data['y'])
